smartsim/log.py

```diff
@@ -123,9 +243,78 @@ def log_to_file(filename: str, log_level: str = "debug") -> None:
                       to allow the file to store more or less verbose
                       logging information.
     :type log_level: int | str
+    :return: strategy function
+    :rtype: callable
     """
-    logger = logging.getLogger("SmartSim")
-    stream = open(  # pylint: disable=consider-using-with
-        filename, "w+", encoding="utf-8"
-    )
-    coloredlogs.install(stream=stream, logger=logger, level=log_level)
+    if logger is None:
+        logger = logging.getLogger("SmartSim")
+
+    # ensure logs are written even if specified dir doesn't exist
+    log_path = pathlib.Path(filename)
+    if not log_path.parent.exists():
+        log_path.parent.mkdir(parents=True, exist_ok=True)
+
+    handler = logging.FileHandler(filename, mode="a+", encoding="utf-8")
+
+    if log_filter:
+        handler.addFilter(log_filter)
+
+    fmt = fmt or EXPERIMENT_LOG_FORMAT
+    formatter = logging.Formatter(fmt=fmt, datefmt=DEFAULT_DATE_FORMAT)
+
+    handler.setFormatter(formatter)
+    handler.setLevel(log_level.upper())
+
+    logger.addHandler(handler)
+    return handler
+
+
+def method_contextualizer(
+    ctx_var: ContextVar[_ContextT],
+    ctx_map: t.Callable[[_T], _ContextT],
+) -> t.Callable[
+    [t.Callable[Concatenate[_T, _PR], _RT]],
+    t.Callable[Concatenate[_T, _PR], _RT],
+]:
+    """Parameterized-decorator factory that enables a target value
+    to be placed into global context prior to execution of the
+    decorated method.
+    Usage Note: the use of `self` below requires that the decorated function is passed
+    the object containing a value that will be modified in the context. `ctx_map`
+    must accept an instance of matching type.
+    :param ctx_var: The ContextVar that will be modified
+    :type ctx_var: ContextVar
+    :param ctx_map: A function that returns the value to be set to ctx_var
+    :type ctx_map: t.Callable[[_T], _ContextT]"""
+
+    def _contextualize(
+        fn: t.Callable[Concatenate[_T, _PR], _RT], /
+    ) -> t.Callable[Concatenate[_T, _PR], _RT]:
+        """Executes the decorated method in a cloned context and ensures
+        `ctx_var` is updated to the value returned by `ctx_map` prior to
+        calling the decorated method"""
+
+        @functools.wraps(fn)
+        def _contextual(
+            self: _T,
+            *args: _PR.args,
+            **kwargs: _PR.kwargs,
+        ) -> _RT:
+            """A decorator operator that runs the decorated method in a new
+            context with the desired contextual information modified."""
+
+            def _ctx_modifier() -> _RT:
+                """Helper to simplify calling the target method with the
+                modified value set in `ctx_var`"""
+                ctx_val = ctx_map(self)
+                token = ctx_var.set(ctx_val)
+                result = fn(self, *args, **kwargs)
+                ctx_var.reset(token)
+                return result
```

Comment:
The context is blown away as soon as the method called during `ctx.run(_ctx_modifier)` returns. This unit test demonstrates:

![image](https://github.com/CrayLabs/SmartSim/assets/3595025/2e351708-805d-4baa-9a79-2ecae00f7874)

I believe I don't have to reset at all because of my basic usage but doing so was safer.